rust-runtime/aws-smithy-runtime/src/client/retries/strategy/standard.rs

```diff
@@ -218,4 +290,302 @@ mod tests {
+        assert_eq!(dur, Duration::from_secs(1));
+        assert_eq!(token_bucket.available_permits(), 495);
+
+        cfg.put(RequestAttempts::new(2));
+        let should_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        let dur = should_retry.expect_delay();
+        assert_eq!(dur, Duration::from_secs(2));
+        assert_eq!(token_bucket.available_permits(), 490);
+
+        ctx.set_output_or_error(Ok(TypeErasedBox::doesnt_matter()));
+
+        cfg.put(RequestAttempts::new(3));
+        let no_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        assert_eq!(no_retry, ShouldAttempt::No);
+        assert_eq!(token_bucket.available_permits(), 495);
+    }
+
+    #[cfg(feature = "test-util")]
+    #[test]
+    fn no_more_attempts() {
+        let (mut cfg, ctx) = setup_test(vec![RetryReason::Error(ErrorKind::ServerError)]);
+        let strategy = StandardRetryStrategy::default()
+            .with_base(|| 1.0)
+            .with_max_attempts(3);
+        cfg.put(StandardTokenBucket::default());
+        let token_bucket = cfg.get::<StandardTokenBucket>().unwrap().clone();
+
+        cfg.put(RequestAttempts::new(1));
+        let should_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        let dur = should_retry.expect_delay();
+        assert_eq!(dur, Duration::from_secs(1));
+        assert_eq!(token_bucket.available_permits(), 495);
+
+        cfg.put(RequestAttempts::new(2));
+        let should_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        let dur = should_retry.expect_delay();
+        assert_eq!(dur, Duration::from_secs(2));
+        assert_eq!(token_bucket.available_permits(), 490);
+
+        cfg.put(RequestAttempts::new(3));
+        let no_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        assert_eq!(no_retry, ShouldAttempt::No);
+        assert_eq!(token_bucket.available_permits(), 490);
+    }
+
+    #[cfg(feature = "test-util")]
+    #[test]
+    fn no_quota() {
+        let (mut cfg, ctx) = setup_test(vec![RetryReason::Error(ErrorKind::ServerError)]);
+        let strategy = StandardRetryStrategy::default()
+            .with_base(|| 1.0)
+            .with_max_attempts(5);
+        cfg.put(StandardTokenBucket::new(5));
+        let token_bucket = cfg.get::<StandardTokenBucket>().unwrap().clone();
+
+        cfg.put(RequestAttempts::new(1));
+        let should_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        let dur = should_retry.expect_delay();
+        assert_eq!(dur, Duration::from_secs(1));
+        assert_eq!(token_bucket.available_permits(), 0);
+
+        cfg.put(RequestAttempts::new(2));
+        let no_retry = strategy.should_attempt_retry(&ctx, &cfg).unwrap();
+        assert_eq!(no_retry, ShouldAttempt::No);
+        assert_eq!(token_bucket.available_permits(), 0);
+    }
+
+    #[cfg(feature = "test-util")]
+    #[test]
+    fn quota_replenishes_on_success() {
```

Comment:
Can we add a test that checks that a successful request with no retries adds tokens to the bucket?

Reply:
done